src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -362,95 +364,118 @@ MultiPlanTree(Query *queryTree)
 static void
 ErrorIfQueryNotSupported(Query *queryTree)
 {
-	char *errorDetail = NULL;
+	char *errorMessage = NULL;
 	bool hasTablesample = false;
 	bool hasUnsupportedJoin = false;
 	bool hasComplexJoinOrder = false;
 	bool hasComplexRangeTableType = false;
 	bool preconditionsSatisfied = true;
+	const char *errorHint = NULL;
+	const char *joinHint = "Consider joining tables on partition column and have "
+						   "equal filter on joining columns.";
+	const char *filterHint = "Consider using an equality filter on the distributed "
+							 "table's partition column.";
 
 	if (queryTree->hasSubLinks)
 	{
 		preconditionsSatisfied = false;
-		errorDetail = "Subqueries other than in from-clause are currently unsupported";
+		errorMessage = "could not run distributed query with subquery outside the "
+					   "FROM clause";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasWindowFuncs)
 	{
 		preconditionsSatisfied = false;
-		errorDetail = "Window functions are currently unsupported";
+		errorMessage = "could not run distributed query with window functions";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->setOperations)
 	{
 		preconditionsSatisfied = false;
-		errorDetail = "Union, Intersect, or Except are currently unsupported";
+		errorMessage = "could not run distributed query with UNION, INTERSECT, or "
+					   "EXCEPT";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->hasRecursive)
 	{
 		preconditionsSatisfied = false;
-		errorDetail = "Recursive queries are currently unsupported";
+		errorMessage = "could not run distributed query with RECURSIVE";
+		errorHint = filterHint;
 	}
 
 	if (queryTree->cteList)
 	{
 		preconditionsSatisfied = false;
-		errorDetail = "Common Table Expressions are currently unsupported";
+		errorMessage = "could not run distributed query with common table expressions";
+		errorHint = filterHint;
```

Comment:
We prune shards using all the restrictions on tables. Postgres provides these via a hook in a query agnostic way. If we hit a single shard in each table and all the shards are on the same node, then we just forward the query with the table names replaced by shard names.